File: js/tech_agent.js
/**
 * NovaTech - AI Emerging Technology Agent & Gemini 3.7 Engine
 */

const TechAgent = {
  // Generate Morning Frontier Briefing
  async generateDailyBriefing() {
    const apiKey = StorageManager.getApiKey();
    const model = StorageManager.getModel();

    if (apiKey) {
      try {
        const prompt = `
You are NovaTech, an elite Principal Research Analyst specializing in emerging and frontier technologies.
Generate a concise, high-impact Daily Emerging Tech Intelligence Briefing for today.
Select the top 4 most critical breakthroughs across AI/Hardware, Quantum, Biotech/Longevity, and Fusion/Clean Energy.

Format as clean JSON:
{
  "date": "Today",
  "title": "Frontier Intelligence Briefing",
  "highlights": [
    {
      "tech": "Technology Name",
      "domain": "Domain Name",
      "summary": "1-2 punchy sentences on the exact breakthrough and commercial consequence.",
      "impact": "High / Critical",
      "readiness": "TRL 6-8"
    }
  ],
  "macroAnalysis": "3-sentence executive summary connecting these breakthroughs to global supply chain and industrial transformation."
}
`;
        const res = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json', temperature: 0.2 }
          })
        });

        if (res.ok) {
          const data = await res.json();
          const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
          if (text) {
            return JSON.parse(text.replace(/^```json\s*/, '').replace(/\s*```$/, '').trim());
          }
        }
      } catch (e) {
        console.warn('Gemini briefing generation fallback:', e);
      }
    }

    // High-Fidelity Curated Daily Briefing
    return {
      date: new Date().toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' }),
      title: 'Frontier Tech Intelligence Digest',
      highlights: [
        {
          tech: 'Silicon Photonic AI Accelerators',
          domain: 'AI & Systems',
          summary: 'Optical matrix multiplication achieves 100x lower energy consumption over traditional GPUs, bypassing electrical interconnect thermal limits.',
          impact: 'Critical',
          readiness: 'TRL 6 (Pilot Trials)'
        },
        {
          tech: 'High-Density Brain-Computer Interfaces (BCI)',
          domain: 'Biotech & Neural Interfaces',
          summary: '1,000+ channel wireless cortical implants successfully restore full desktop digital control and sub-50ms robotic arm dexterity.',
          impact: 'Critical',
          readiness: 'TRL 7 (Human Trials)'
        },
        {
          tech: 'Compact HTS Magnetic Fusion',
          domain: 'Energy & Climate',
          summary: '20-Tesla REBCO superconducting magnets demonstrate stable plasma confinement, accelerating commercial net-electricity timeline to 2030.',
          impact: 'Transformational',
          readiness: 'TRL 5 (Prototype)'
        },
        {
          tech: 'Fault-Tolerant Logical Qubit Scaling',
          domain: 'Quantum & Hardware',
          summary: 'Demonstration of 100+ error-corrected logical qubits on neutral-atom optical tweezers paving way for quantum chemistry simulations.',
          impact: 'High',
          readiness: 'TRL 5 (Lab Validation)'
        }
      ],
      macroAnalysis: 'The convergence of photonic interconnects and high-density neural telemetry signals a shift toward physical-digital computing architectures that operate beyond traditional Von Neumann silicon bottlenecks.'
    };
  },

  // Deep Dive Synthesis for a Specific Technology
  async getDeepDive(techId) {
    const tech = TechKnowledgeBase.getTechById(techId);
    if (!tech) return null;

    const apiKey = StorageManager.getApiKey();
    const model = StorageManager.getModel();

    if (apiKey) {
      try {
        const prompt = `
You are NovaTech Principal Analyst. Provide an exhaustive technical briefing on the emerging technology: "${tech.name}".
Domain: ${tech.domain} | TRL: ${tech.trl} | Disruption Score: ${tech.impactScore}/100.

Return JSON:
{
  "executiveSummary": "2-3 comprehensive sentences",
  "underlyingMechanism": "Deep technical explanation of how the breakthrough operates at a physics/hardware/software level",
  "competitiveLandscape": [
    { "player": "Company / Lab name", "focus": "Their specific architectural approach or advantage" }
  ],
  "commercialTimeline": "Detailed 3-5 year projection",
  "technicalBottlenecks": ["Bullet 1", "Bullet 2", "Bullet 3"],
  "industrialDisruption": "Which legacy markets and industries will be disrupted"
}
`;
        const res = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json', temperature: 0.2 }
          })
        });

        if (res.ok) {
          const data = await res.json();
          const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
          if (text) {
            const parsed = JSON.parse(text.replace(/^```json\s*/, '').replace(/\s*```$/, '').trim());
            return { ...tech, ...parsed };
          }
        }
      } catch (e) {
        console.warn('Gemini deep dive fallback:', e);
      }
    }

    // Curated Deep Dive Fallback
    return {
      ...tech,
      executiveSummary: tech.summary,
      underlyingMechanism: tech.breakthrough,
      competitiveLandscape: tech.leadingPlayers.map(p => ({
        player: p,
        focus: `Frontier research and active deployment in ${tech.domain}.`
      })),
      commercialTimeline: `Expected commercial scaling between ${tech.timeline}, reaching main enterprise adoption at TRL 8–9.`,
      technicalBottlenecks: tech.bottlenecks,
      industrialDisruption: `Major disruption anticipated in traditional ${tech.domain} workflows, dropping compute/energy barriers and enabling new market categories.`
    };
  },

  // Conversational Frontier Tech Analyst
  async askAnalyst(question, contextTech = null) {
    const apiKey = StorageManager.getApiKey();
    const model = StorageManager.getModel();

    if (apiKey) {
      try {
        const prompt = `
You are NovaTech, an authoritative, deeply knowledgeable Principal Frontier Technology Analyst.
${contextTech ? `Current Context Technology: "${contextTech.name}" (${contextTech.domain}, TRL ${contextTech.trl})` : ''}

User Query: "${question}"

Provide a concise, technically rigorous, and objective breakdown with bullet points, naming relevant startups, silicon architectures, or scientific papers.
`;
        const res = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }] })
        });
        if (res.ok) {
          const data = await res.json();
          const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
          if (text) return text;
        }
      } catch (e) {
        console.warn('Gemini analyst chat fallback:', e);
      }
    }

    // Intelligent Offline Analyst Logic
    const q = question.toLowerCase();

    if (q.includes('photonic') || q.includes('optical') || q.includes('light')) {
      return `💡 **Silicon Photonics vs GPU Accelerators:**\n\n- **Core Physics:** Photonic chips perform matrix multiplication using passive light interference through Mach-Zehnder mesh arrays, computing at the speed of light.\n- **Energy Efficiency:** Consumes ~1 femtojoule per MAC operation (versus ~10 picojoules on 3nm digital CMOS)—a 100x improvement.\n- **Leading Players:** Lightmatter, Celestial AI, Luminous Computing.\n- **Bottleneck:** Thermal drift of laser wavelengths and conversion back to electrical signals for memory storage.`;
    }

    if (q.includes('fusion') || q.includes('clean energy') || q.includes('hts')) {
      return `⚡ **High-Temperature Superconducting (HTS) Fusion Status:**\n\n- **The Breakthrough:** REBCO (Rare-Earth Barium Copper Oxide) tapes allow magnetic fields exceeding 20 Tesla, shrinking the required tokamak volume by 40x.\n- **Net Energy (Q > 1):** Commonwealth Fusion Systems' SPARC reactor and Helion Energy's Polaris are targeting commercial pilot demos between 2026–2029.\n- **Primary Challenge:** Tritium self-sufficiency breeding blankets and high neutron flux wall durability.`;
    }

    if (q.includes('quantum') || q.includes('qubit')) {
      return `⚛️ **Quantum Computing Roadmap (2026–2030):**\n\n- **Shift to Logical Qubits:** The industry has moved beyond noisy physical qubit counts to fault-tolerant logical qubits using surface and color codes.\n- **Modality Race:** Neutral atoms (QuEra) and trapped ions (Quantinuum) are leading in error mitigation, while superconducting transmons (IBM/Google) lead in raw gate speed.\n- **PQC Transition:** NIST has finalized post-quantum encryption standards (ML-KEM, ML-DSA) as quantum decryption timelines shorten.`;
    }

    return `🔬 **Frontier Intelligence Assessment:**\n\nBased on current technological velocity across ${contextTech ? contextTech.domain : 'frontier domains'}, major industrial inflection points are occurring in photonic interconnects, high-density neural telemetry, and generative enzyme engineering. You can ask me to compare specific architectures, evaluate startup roadmaps, or explain underlying physics anytime!`;
  },

  // Format and Launch WhatsApp Briefing Dispatch
  dispatchWhatsAppBriefing(briefing, recipientPhone) {
    const cleanPhone = (recipientPhone || '+919876543210').replace(/[^0-9]/g, '');

    const lines = [
      `⚡ *NovaTech Frontier Emerging Tech Briefing*`,
      `📅 *Date:* ${briefing.date || 'Today'}`,
      `---------------------------------------`,
      `*TOP BREAKTHROUGHS TO WATCH:*`
    ];

    briefing.highlights?.forEach((h, i) => {
      lines.push(
        `\n🔹 *${i + 1}. ${h.tech}* (${h.domain})`,
        `   • *Status:* ${h.readiness} | *Impact:* ${h.impact}`,
        `   • *Summary:* ${h.summary}`
      );
    });

    lines.push(
      `\n---------------------------------------`,
      `📊 *Macro Executive Takeaway:*`,
      `${briefing.macroAnalysis || 'Frontier computing and biotechnology breakthroughs are accelerating towards enterprise adoption.'}`,
      `\n_Sent via NovaTech Autonomous Emerging Tech Intelligence Agent_`
    );

    const fullMessage = lines.join('\n');
    const waUrl = `https://wa.me/${cleanPhone}?text=${encodeURIComponent(fullMessage)}`;
    window.open(waUrl, '_blank');
  }
};

File: python_agent/tech_agent.py
#!/usr/bin/env python3
"""
NovaTech - Autonomous Emerging Tech Intelligence Agent (Python CLI & SDK)
Provides frontier tech radar monitoring, daily digests, and deep-dive technical research.
"""

import os
import sys
import json
import argparse
from typing import Dict, Any, List, Optional

try:
    from google import genai
    HAS_GENAI = True
except ImportError:
    HAS_GENAI = False


class EmergingTechAgent:
    """Autonomous Emerging Technology Research & Synthesis Agent."""

    CURATED_TECH = {
        "photonic_ai": {
            "name": "Silicon Photonic AI Accelerators",
            "domain": "AI & Systems",
            "trl": 6,
            "impact": 94,
            "timeline": "2026–2028",
            "summary": "Optical matrix multiplication processors using photons rather than electrons, cutting energy per MAC operation by 100x.",
            "players": ["Lightmatter", "Celestial AI", "Luminous Computing", "Intel Labs"],
            "bottlenecks": "Laser thermal drift and optical-to-electrical signal conversion."
        },
        "hts_fusion": {
            "name": "High-Temperature Superconducting (HTS) Compact Fusion",
            "domain": "Energy & Materials",
            "trl": 5,
            "impact": 100,
            "timeline": "2028–2033",
            "summary": "20+ Tesla REBCO superconducting magnets enabling compact tokamak reactors targeting commercial net electricity (Q > 5).",
            "players": ["Commonwealth Fusion Systems", "Helion Energy", "Tokamak Energy", "TAE"],
            "bottlenecks": "Tritium breeding efficiency and plasma-facing tungsten wall degradation."
        },
        "bci_neural": {
            "name": "High-Density Brain-Computer Interfaces (BCI)",
            "domain": "Biotech & Neural",
            "trl": 7,
            "impact": 97,
            "timeline": "2025–2028",
            "summary": "1,000+ channel wireless implantable cortical telemetry streaming motor intent to digital devices with sub-50ms latency.",
            "players": ["Neuralink", "Synchron", "Precision Neuroscience", "Blackrock Neurotech"],
            "bottlenecks": "Long-term glial scar encapsulation over 5+ years."
        },
        "logical_qubits": {
            "name": "Fault-Tolerant Logical Qubit Processors",
            "domain": "Quantum Computing",
            "trl": 5,
            "impact": 99,
            "timeline": "2027–2031",
            "summary": "Quantum processors using active surface/color code error correction to achieve sub-10^-6 physical gate error rates.",
            "players": ["QuEra Computing", "IBM Quantum", "Google Quantum AI", "Quantinuum"],
            "bottlenecks": "Cryogenic refrigeration scaling and optical control laser phase noise."
        }
    }

    def __init__(self, api_key: Optional[str] = None, model: str = "gemini-3.7-flash"):
        self.api_key = api_key or os.environ.get("GEMINI_API_KEY")
        self.model = model
        self.client = None
        if self.api_key and HAS_GENAI:
            try:
                self.client = genai.Client(api_key=self.api_key)
            except Exception:
                pass

    def get_daily_briefing(self) -> Dict[str, Any]:
        """Generates a structured daily frontier technology briefing."""
        if self.client:
            try:
                prompt = """
You are NovaTech Principal Research Analyst. Generate a concise Daily Emerging Technology Briefing in JSON:
{
  "date": "Today",
  "title": "Frontier Tech Intelligence Digest",
  "highlights": [
    { "tech": "Technology Name", "domain": "Domain", "summary": "1-2 sentence breakthrough impact", "trl": 6 }
  ],
  "macroAnalysis": "2-sentence executive trend"
}
"""
                resp = self.client.models.generate_content(model=self.model, contents=prompt)
                if resp and resp.text:
                    text = resp.text.strip().replace("```json", "").replace("```", "").strip()
                    return json.loads(text)
            except Exception:
                pass

        # Fallback
        return {
            "date": "September 2026",
            "title": "Frontier Technology Intelligence Briefing",
            "highlights": [
                {
                    "tech": "Silicon Photonic AI Accelerators",
                    "domain": "AI & Systems",
                    "summary": "Optical matrix multiplication delivers 100x energy efficiency improvement over 3nm digital CMOS chips.",
                    "trl": 6
                },
                {
                    "tech": "High-Density Neural Telemetry (BCI)",
                    "domain": "Biotech & Neural",
                    "summary": "1,000+ channel wireless implants restore full robotic arm dexterity and digital typing speeds.",
                    "trl": 7
                },
                {
                    "tech": "20-Tesla REBCO Compact Fusion",
                    "domain": "Energy & Materials",
                    "summary": "Superconducting magnet arrays shrink reactor volume 40x, targeting grid pilot tests by 2029.",
                    "trl": 5
                }
            ],
            "macroAnalysis": "Breakthroughs in photonic interconnects and high-density neural telemetry signal a shift toward physical-digital computing architectures operating beyond Von Neumann limits."
        }

    def deep_dive(self, tech_name: str) -> Dict[str, Any]:
        """Generates an in-depth architectural and market analysis of a technology."""
        if self.client:
            try:
                prompt = f"""
Provide an exhaustive technical briefing on the emerging technology: "{tech_name}".
Return JSON:
{{
  "name": "{tech_name}",
  "summary": "Executive summary",
  "mechanism": "Deep technical physics/software explanation",
  "leadingPlayers": ["Company 1", "Company 2"],
  "timeline": "2026-2029",
  "bottlenecks": ["Bottleneck 1", "Bottleneck 2"],
  "marketDisruption": "Affected legacy markets"
}}
"""
                resp = self.client.models.generate_content(model=self.model, contents=prompt)
                if resp and resp.text:
                    text = resp.text.strip().replace("```json", "").replace("```", "").strip()
                    return json.loads(text)
            except Exception:
                pass

        # Fallback search
        norm = tech_name.lower()
        match = next((v for k, v in self.CURATED_TECH.items() if k in norm or v["name"].lower() in norm), None)
        if match:
            return {
                "name": match["name"],
                "summary": match["summary"],
                "mechanism": f"Advanced {match['domain']} breakthrough operating at TRL {match['trl']}.",
                "leadingPlayers": match["players"],
                "timeline": match["timeline"],
                "bottlenecks": [match["bottlenecks"]],
                "marketDisruption": "High disruption across legacy computing, energy, or healthcare systems."
            }

        return {
            "name": tech_name,
            "summary": f"Frontier breakthrough in {tech_name}.",
            "mechanism": "Novel physical and architectural implementation accelerating performance.",
            "leadingPlayers": ["Frontier Startups & Research Labs"],
            "timeline": "2026–2030",
            "bottlenecks": ["Manufacturing yield", "Scaling cost"],
            "marketDisruption": "Direct challenge to incumbent market standards."
        }

    def format_markdown(self, data: Dict[str, Any], is_briefing: bool = True) -> str:
        if is_briefing:
            md = [
                f"# ⚡ NovaTech Frontier Technology Briefing",
                f"> **Date:** {data.get('date', 'Today')}",
                f"\n### 📊 Top Breakthrough Highlights\n"
            ]
            for h in data.get("highlights", []):
                md.append(f"- **{h.get('tech')}** (`{h.get('domain')}` | TRL {h.get('trl', 6)})")
                md.append(f"  *{h.get('summary')}*\n")

            md.append(f"### 🌐 Macro Analysis")
            md.append(f"{data.get('macroAnalysis', '')}\n")
            md.append("*Generated by NovaTech Emerging Tech Intelligence Agent*")
            return "\n".join(md)
        else:
            md = [
                f"# 🔬 NovaTech Deep-Dive: {data.get('name')}",
                f"> **Executive Summary:** {data.get('summary')}",
                f"\n### ⚙️ Core Technical Mechanism",
                f"{data.get('mechanism')}",
                f"\n### 🏢 Key Commercial & Research Leaders",
                f"- " + "\n- ".join(data.get("leadingPlayers", [])),
                f"\n### ⚠️ Critical Bottlenecks",
                f"- " + "\n- ".join(data.get("bottlenecks", [])),
                f"\n### 🚀 Market Disruption Horizon: {data.get('timeline')}",
                f"{data.get('marketDisruption')}\n"
            ]
            return "\n".join(md)


def main():
    parser = argparse.ArgumentParser(description="NovaTech Emerging Technology Intelligence CLI")
    parser.add_argument("--briefing", "-b", action="store_true", help="Generate today's frontier tech briefing")
    parser.add_argument("--tech", "-t", type=str, help="Deep dive into a specific emerging technology")
    parser.add_argument("--export-md", type=str, help="Output Markdown report path")
    parser.add_argument("--export-json", type=str, help="Output JSON path")

    args = parser.parse_args()
    agent = EmergingTechAgent()

    if args.tech:
        print(f"🔬 Synthesizing deep dive for: {args.tech}...")
        report = agent.deep_dive(args.tech)
        md_text = agent.format_markdown(report, is_briefing=False)
    else:
        print("⚡ Generating Daily Frontier Tech Briefing...")
        report = agent.get_daily_briefing()
        md_text = agent.format_markdown(report, is_briefing=True)

    print("\n" + "=" * 60)
    print(md_text)
    print("=" * 60 + "\n")

    if args.export_md:
        with open(args.export_md, "w", encoding="utf-8") as f:
            f.write(md_text)
        print(f"✅ Saved Markdown to {args.export_md}")

    if args.export_json:
        with open(args.export_json, "w", encoding="utf-8") as f:
            json.dump(report, f, indent=2)
        print(f"✅ Saved JSON to {args.export_json}")


if __name__ == "__main__":
    main()
